contracts/src/UniswapV3Sampler.sol

```diff
@@ -165,39 +106,25 @@ contract UniswapV3Sampler {
         uniswapPaths = new bytes[](makerTokenAmounts.length);
         uniswapGasUsed = new uint256[](makerTokenAmounts.length);
 
-        for (uint256 i = 0; i < makerTokenAmounts.length; ++i) {
-            // Pick the best result from the pool paths.
-            uint256 topSellAmount = 0;
-            for (uint256 j = 0; j < poolPaths.length; ++j) {
-                if (!isValidPoolPath(poolPaths[j])) {
-                    continue;
-                }
-
-                // quoter requires path to be reversed for buys.
-                bytes memory uniswapPath = _toUniswapPath(reversedPath, poolPaths[j]);
-                try quoter.quoteExactOutput{gas: QUOTE_GAS}(uniswapPath, makerTokenAmounts[i]) returns (
-                    uint256 sellAmount,
-                    uint160[] memory /* sqrtPriceX96AfterList */,
-                    uint32[] memory /* initializedTicksCrossedList */,
-                    uint256 gasUsed
-                ) {
-                    if (topSellAmount == 0 || topSellAmount >= sellAmount) {
-                        topSellAmount = sellAmount;
-                        // But the output path should still be encoded for sells.
-                        uniswapPaths[i] = _toUniswapPath(path, _reversePoolPath(poolPaths[j]));
-                        uniswapGasUsed[i] = gasUsed;
-                    }
-                } catch {}
+        for (uint256 i = 0; i < poolPaths.length; ++i) {
+            if (!isValidPoolPath(poolPaths[i])) {
+                continue;
             }
-            // Break early if we can't complete the buys.
-            if (topSellAmount == 0) {
-                // HACK(kimpers): To avoid too many local variables, paths and gas used is set directly in the loop
-                // then reset if no valid valid quote was found
-                uniswapPaths[i] = "";
-                uniswapGasUsed[i] = 0;
-                break;
+
+            bytes memory uniswapPath = toUniswapPath(reversedPath, poolPaths[i]);
+            (uint256[] memory amountsIn, uint256[] memory gasEstimate) = multiQuoter.quoteExactMultiOutput(
+                quoter.factory(),
+                uniswapPath,
+                makerTokenAmounts
+            );
+
+            for (uint256 j = 0; j < amountsIn.length; ++j) {
+                if (amountsIn[j] > 0 && (takerTokenAmounts[j] == 0 || takerTokenAmounts[j] > amountsIn[j])) {
```

Comment:
nit/optional

Is this condition equivalent to `takerTokenAmounts[j] < amountsIn[j]` ?

Reply:
Not quite, we want the best buy sample, so the pool that results in the smallest `amountsIn[j]` to be passed in. However, `amountsIn[j]` can be 0 so we have to filter out the 0s.